### PR TITLE
auth: Fix data race in Upsert

### LINKED ISSUE #33899: [DATA RACE]: github.com/cilium/cilium/operator/auth/spire.(*Client).Upsert()
```
2024-07-18T14:07:45.630548802Z Read at 0x00c00029d0f0 by goroutine 290:
2024-07-18T14:07:45.630714080Z   github.com/cilium/cilium/operator/auth/spire.(*Client).Upsert()
2024-07-18T14:07:45.630723558Z       /go/src/github.com/cilium/cilium/operator/auth/spire/client.go:194 +0x6f
2024-07-18T14:07:45.630850324Z   github.com/cilium/cilium/operator/auth.(*IdentityWatcher).run()
2024-07-18T14:07:45.630859782Z       /go/src/github.com/cilium/cilium/operator/auth/watcher.go:70 +0x485
2024-07-18T14:07:45.630982661Z   github.com/cilium/cilium/operator/auth.registerIdentityWatcher.func1.1()
2024-07-18T14:07:45.630991528Z       /go/src/github.com/cilium/cilium/operator/auth/watcher.go:56 +0x3e
2024-07-18T14:07:45.631090082Z   github.com/cilium/workerpool.(*WorkerPool).run.func1()
2024-07-18T14:07:45.631099198Z       /go/src/github.com/cilium/cilium/vendor/github.com/cilium/workerpool/workerpool.go:181 +0xe1
2024-07-18T14:07:45.631220284Z 
2024-07-18T14:07:45.631433492Z Previous write at 0x00c00029d0f0 by goroutine 266:
2024-07-18T14:07:45.631515585Z   github.com/cilium/cilium/operator/auth/spire.(*Client).onStart.func1()
2024-07-18T14:07:45.631523961Z       /go/src/github.com/cilium/cilium/operator/auth/spire/client.go:134 +0x269
2024-07-18T14:07:45.632197738Z 
2024-07-18T14:07:45.632210491Z Goroutine 290 (running) created at:
2024-07-18T14:07:45.632215411Z   github.com/cilium/workerpool.(*WorkerPool).run()
2024-07-18T14:07:45.632220270Z       /go/src/github.com/cilium/cilium/vendor/github.com/cilium/workerpool/workerpool.go:178 +0x76
2024-07-18T14:07:45.632257780Z   github.com/cilium/workerpool.NewWithContext.gowrap1()
2024-07-18T14:07:45.632262679Z       /go/src/github.com/cilium/cilium/vendor/github.com/cilium/workerpool/workerpool.go:68 +0x4f
2024-07-18T14:07:45.632267698Z 
2024-07-18T14:07:45.632286052Z Goroutine 266 (finished) created at:
2024-07-18T14:07:45.632291052Z   github.com/cilium/cilium/operator/auth/spire.(*Client).onStart()
2024-07-18T14:07:45.632295771Z       /go/src/github.com/cilium/cilium/operator/auth/spire/client.go:126 +0x8d
2024-07-18T14:07:45.632300409Z   github.com/cilium/cilium/operator/auth/spire.(*Client).onStart-fm()
2024-07-18T14:07:45.632305078Z       <autogenerated>:1 +0x47
2024-07-18T14:07:45.632309797Z   github.com/cilium/hive/cell.Hook.Start()
2024-07-18T14:07:45.632314636Z       /go/src/github.com/cilium/cilium/vendor/github.com/cilium/hive/cell/lifecycle.go:42 +0x70
2024-07-18T14:07:45.632318894Z   github.com/cilium/hive/cell.(*Hook).Start()
2024-07-18T14:07:45.632323292Z       <autogenerated>:1 +0x1f
2024-07-18T14:07:45.632327710Z   github.com/cilium/hive/cell.(*DefaultLifecycle).Start()
2024-07-18T14:07:45.632332189Z       /go/src/github.com/cilium/cilium/vendor/github.com/cilium/hive/cell/lifecycle.go:107 +0x46a
2024-07-18T14:07:45.632336527Z   github.com/cilium/cilium/operator/cmd.runOperator.func1()
2024-07-18T14:07:45.632340774Z       /go/src/github.com/cilium/cilium/operator/cmd/root.go:432 +0x7e
2024-07-18T14:07:45.632345373Z   k8s.io/client-go/tools/leaderelection.(*LeaderElector).Run.gowrap2()
2024-07-18T14:07:45.632350011Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:213 +0x4f
```
seen with https://github.com/cilium/cilium/commit/54796b085eb2f8c695db74cef95521e732effacf

### PR DESCRIPTION
**Pull request by chaunceyjiang**
Fixes: #33899

```release-note
auth: Fix data race in Upsert
```

### CODE REVIEW

**Review by pippolo84 — COMMENTED** [review]
The race seems to be caused by calling `*Client.Upsert` while the internal client (that is, `*Client.entry`) is being written in the `onStart` method. I don't think that the sync.Once will be of any help, since the assignment it is guarding:

```go
c.entry = entryv1.NewEntryClient(conn)
```

is already executed at most once (it is part of a start hook and as soon as the connection succeeds, we break from the loop).

Instead, I think we should add a lock to guard the `entry` field, and take it before:

- writing it in the Start hook
- reading it in the Update, Delete and List

so that we won't race to access `entry` from the start hook goroutine and the workerpool one.

File: operator/auth/spire/client.go [hunk 1]
```diff
@@ -287,6 +296,8 @@ func (c *Client) List(ctx context.Context) ([]string, error) {
 // listEntries returns the list of entries for the given ID.
 // The maximum number of entries returned is 1, so page token can be ignored.
 func (c *Client) listEntries(ctx context.Context, id string) (*entryv1.ListEntriesResponse, error) {
+	c.entryMutex.RLock()
+	defer c.entryMutex.RUnlock()
```
Comment (pippolo84) [inline operator/auth/spire/client.go]: `listEntries` is called from `Upsert` and `Delete` only, where the read-lock is already taken, so we don't need this.

**Review by pippolo84 — CHANGES_REQUESTED** [review]

File: operator/auth/spire/client.go (hunk 1 above)
Comment (chaunceyjiang) [reply operator/auth/spire/client.go]: Updated.

**Review by pippolo84 — APPROVED** [review]
Thanks! :100:

**Pull request merged**